RecoTauTag/RecoTau/src/PFRecoTauClusterVariables.cc

```diff
@@ -0,0 +1,367 @@
+      }
+    }
+    return LeadingTracknormalizedChi2;
+  }
+  /// return ratio of energy in ECAL over sum of energy in ECAL and HCAL
+  float tau_Eratio(const reco::PFTau& tau) {
+    std::vector<reco::PFCandidatePtr> constsignal = tau.signalPFCands();
+    float EcalEnInSignalPFCands = 0;
+    float HcalEnInSignalPFCands = 0;
+    typedef std::vector <reco::PFCandidatePtr>::iterator constituents_iterator;
+    for(constituents_iterator it=constsignal.begin(); it != constsignal.end(); ++it) {
+      reco::PFCandidatePtr & icand = *it;
+      EcalEnInSignalPFCands += icand -> ecalEnergy();
+      HcalEnInSignalPFCands += icand -> hcalEnergy();
+    }
+    float total = EcalEnInSignalPFCands + HcalEnInSignalPFCands;
+    if(total==0){ 
+      return -1;
+    }
+    return EcalEnInSignalPFCands/total;
+  }
+  float tau_Eratio(const pat::Tau& tau) {
+    float EcalEnInSignalCands = tau.ecalEnergy();
+    float HcalEnInSignalCands = tau.hcalEnergy();
+    float total = EcalEnInSignalCands + HcalEnInSignalCands;
+    if(total == 0){ 
+      return -1;
+    }
+    return EcalEnInSignalCands/total;
+  }
+  /// return sum of pt weighted values of distance to tau candidate for all pf photon candidates, 
+  /// which are associated to signal; depending on var the distance is in 0=:dr, 1=:deta, 2=:dphi 
+  float pt_weighted_dx(const reco::PFTau& tau, int mode, int var, int decaymode) {  
+    float sum_pt = 0.;
+    float sum_dx_pt = 0.;
+    float signalrad = std::max(0.05, std::min(0.1, 3./std::max(1., tau.pt())));
+    int is3prong = (decaymode==10);
+    const auto& cands = getPFGammas(tau, mode < 2);
+    for (const auto& cand : cands) {
+      // only look at electrons/photons with pT > 0.5
+      if ((float)cand->pt() < 0.5){
+        continue;
+      }
+      float dr = reco::deltaR((float)cand->eta(),(float)cand->phi(),(float)tau.eta(),(float)tau.phi());
+      float deta = std::abs((float)cand->eta() - (float)tau.eta());
+      float dphi = std::abs(reco::deltaPhi((float)cand->phi(), (float)tau.phi()));
+      float pt = cand->pt();
+      bool flag = isInside(pt, deta, dphi);
+      if(is3prong==0){
+        if (mode == 2 || (mode == 0 && dr < signalrad) || (mode == 1 && dr > signalrad)) {
+          sum_pt += pt;
+          if (var == 0)
+            sum_dx_pt += pt * dr;
+          else if (var == 1)
+            sum_dx_pt += pt * deta;
+          else if (var == 2)
+            sum_dx_pt += pt * dphi;
+        }
+      }
+      else if(is3prong==1){
+        if( (mode==2 && flag==false) || (mode==1 && flag==true) || mode==0){
+          sum_pt += pt;
+          if (var == 0)
+            sum_dx_pt += pt * dr;
+          else if (var == 1)
+            sum_dx_pt += pt * deta;
+          else if (var == 2)
+            sum_dx_pt += pt * dphi;
+        }
+      }
+    }
+    if (sum_pt > 0.){
+      return sum_dx_pt/sum_pt;  
+    }
+    return 0.;
+  }
+  float pt_weighted_dx(const pat::Tau& tau, int mode, int var, int decaymode) {
```

Comment:
the implementation is identical to ```pt_weighted_dx(const PFTau& tau```. It looks like a good candidate for a template

Reply:
here is one option
https://github.com/cms-sw/cmssw/commit/2169a12485d92c3dba5e405664f6a733518605b0

is this really unreadable?
I think that the danger of missing an update in either of the two duplicated places is worse than slight increase in complexity.